examples/wasi-http-rust-async/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+            );
+
+            let mut sink = outgoing_response_body(wakers, &response);
+
+            ResponseOutparam::set(response_out, Ok(response));
+
+            while let Some((url, result)) = results.next().await {
+                sink.send(Some(
+                    match result {
+                        Ok(hash) => format!("{url}: {hash}\n"),
+                        Err(e) => format!("{url}: {e:?}\n"),
+                    }
+                    .into_bytes(),
+                ))
+                .await?;
+            }
+
+            sink.send(None).await?;
+        }
+
+        (Method::Post, Some("/echo")) => {
+            let response = OutgoingResponse::new(
+                200,
+                Fields::new(
+                    &headers
+                        .iter()
+                        .filter_map(|(k, v)| {
+                            (k == "content-type").then_some((k.clone(), v.clone()))
+                        })
+                        .collect::<Vec<_>>(),
+                ),
+            );
+
+            let mut sink = outgoing_response_body(wakers.clone(), &response);
+
+            ResponseOutparam::set(response_out, Ok(response));
+
+            let mut stream = incoming_request_body(wakers, request);
+            while let Some(chunk) = stream.try_next().await? {
+                sink.send(Some(chunk)).await?;
+            }
+
+            sink.send(None).await?;
+        }
+
+        _ => {
+            let response = OutgoingResponse::new(405, Fields::new(&[]));
+
+            let body = response.write().expect("response should be writable");
+
+            ResponseOutparam::set(response_out, Ok(response));
+
+            OutgoingBody::finish(body, None);
+        }
+    }
+
+    Ok(())
+}
+
+async fn hash(wakers: Wakers, url: Url) -> Result<String> {
+    let request = OutgoingRequest::new(
+        &Method::Get,
+        Some(url.path()),
+        Some(&match url.scheme() {
+            "http" => Scheme::Http,
+            "https" => Scheme::Https,
+            scheme => Scheme::Other(scheme.into()),
+        }),
+        Some(url.authority()),
+        Fields::new(&[]),
+    );
+
+    let response = outgoing_request_send(wakers.clone(), request).await?;
+
+    let status = response.status();
+
+    if !(200..300).contains(&status) {
+        bail!("unexpected status: {status}");
+    }
+
+    let mut body = incoming_response_body(wakers, response);
+
+    let mut hasher = Sha256::new();
+    while let Some(chunk) = body.try_next().await? {
+        hasher.update(&chunk);
+    }
+
+    Ok(hex::encode(hasher.finalize()))
+}
+
+fn outgoing_response_body(
+    wakers: Wakers,
+    response: &OutgoingResponse,
+) -> impl Sink<Option<Vec<u8>>, Error = Error> {
+    outgoing_body(
+        wakers,
+        response.write().expect("response should be writable"),
+    )
+}
+
+fn outgoing_body(wakers: Wakers, body: OutgoingBody) -> impl Sink<Option<Vec<u8>>, Error = Error> {
```

Comment:
Is the plan to move this into some shared library?

Reply:
Yeah, specifically the Spin Rust SDK.

Reply:
I've moved the executor and helper functions into the SDK.  Specifically, they're now generated by the `wasi_http_component` proc macro I've added due to limitations in `wit-bindgen`.  I'm hoping we can port `wasmtime-wit-bindgen`'s `with` directive to `wit-bindgen` in time for Spin 2.0, which will allow us to move that stuff out of the proc macro.